### PR TITLE
feature: add gooey CLI

### PR DESCRIPTION
**Pull request by pollend**
Squashed commit of the following:

commit 89b591f14997ff616eaa83b234b56005654bf1f9
Author: Cervator <cervator@gmail.com>
Date:   Fri Jul 10 22:55:01 2020 -0500

    feat: working recurse command in the new style, misc cleanup and utility

commit 85433d5e3f0f9d513be07d849753f0c5433c5bd3
Author: Cervator <cervator@gmail.com>
Date:   Fri Jul 10 16:17:53 2020 -0500

    chore: rename stuff, add package structure, tweak some comments, more recurse logic

commit 53662ac93738a71c843c4622bddca5004eb33b2d
Author: Cervator <cervator@gmail.com>
Date:   Thu Jul 9 23:23:05 2020 -0500

    chore: adjust the Gradle-based GooeyCLI setup

    * Moves it to a maybe more sensible place in the gradle directory where the wrapper lives, this seems kinda related
    * Standardizes the source path so we don't need to explicitly configure it
    * Drops the Groovy dependency (in Gradle) down to match what's currently in the Gradle Wrapper (although it would still get fetched and used separately for any sort of GooeyCLI *build* process)

commit c5ed6d3866d346a9ee58dc32a16e3ba5d1d12d1d
Author: Michael <polli104@mail.chapman.edu>
Date:   Mon Jul 6 13:14:35 2020 -0700

    correct synopsisSubcommandLabel

commit fc46d278bf36e4bd7c1a3c61b629082b320ccf91
Author: Michael <polli104@mail.chapman.edu>
Date:   Mon Jul 6 12:29:17 2020 -0700

    add sourceset for cli config

commit d9e2e189c419239a7766420345baf1662bab9dd1
Author: Cervator <cervator@gmail.com>
Date:   Sun Jul 5 22:59:08 2020 -0500

    chore: fix path on Windows, avoid some warning about logging itself

commit 4a9dc6447a7f2f31df04c23e6216380298bdf76e
Author: Michael <polli104@mail.chapman.edu>
Date:   Sun Jul 5 20:35:35 2020 -0700

    move CLI and added into package

commit d21e82a420c0d576d6b37e785797b5be1ba1e7d9
Author: Cervator <cervator@gmail.com>
Date:   Sat Jul 4 15:09:45 2020 -0500

    chore: Minor cleanup, plus bring back the module file Git thought was being deleted rather than replaced

commit 05ca0f878e3fa741495ba706904a771477837055
Author: Cervator <cervator@gmail.com>
Date:   Fri Jul 3 20:51:36 2020 -0500

    feat: GooeyCLI - replacement utility scripting system based on PicoCLI.

    Still a proof of concept, with a limited amount of old functionality converted to showcase the new approach. Relies on a groovyw tweak, renames the scripts, and lets PicoCLI worry about all the CLI structuring

### CODE REVIEW

**Comment by pollend** [comment]
I ended up just getting rid of a lot of this. I see that there was some generalization trying to fit Module/Facade and whatever other items we plan to support but without seeing those API and how they would fit into this scheme I don't really see a good reason to introduce this type of generalization. seems to complicate things as we try to square one problem and fit it into another.

**Comment by pollend** [comment]
I was looking at this for formatting the diff but this is ok for the moment: https://github.com/grigoriy-orlov/git-stats/blob/master/src/main/java/ru/ares4322/gitstats/GitStatsImpl.java

File: gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/InitCommand.groovy [hunk 1]
```diff
@@ -30,14 +30,15 @@ class InitCommand extends BaseCommandType implements Runnable {
 
     void run() {
         println Ansi.AUTO.string("@|bold,green,underline Time to initialize $distro !|@")
-        def targetDistroURL = "https://raw.githubusercontent.com/$githubOrg/Index/master/distros/$distro/gradle.properties"
+        String origin = gitOptions.resolveOrigin()
```
Comment (skaldarnar) [inline gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/InitCommand.groovy:31]: How does this resolve? We'd run this command in the `MovingBlocks/Terasology` repository, but the index repo is located in a different org, `Terasology/Index`...

Comment (pollend) [reply gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/InitCommand.groovy:31]: because it's resolving to Terasology?

File: gradle/gooeycli/build.gradle [hunk 2]
```diff
@@ -0,0 +1,39 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+apply plugin: 'groovy'
+apply plugin: 'ivy-publish'
+
+group 'org.terasology.cli'
+
+repositories {
+    jcenter()
```
Comment (DarkWeird) [inline gradle/gooeycli/build.gradle:9]: Jcenter will down after 1 Feb 2022

File: settings.gradle [hunk 3]
```diff
@@ -3,6 +3,7 @@ import groovy.io.FileType
 rootProject.name = 'Terasology'
 
 includeBuild("build-logic")
+includeBuild("gradle/gooeycli")
```
Comment (DarkWeird) [inline settings.gradle:6]: After `.gradle` cleanup - don't want to configure gooeycli..
```
Project directory '/home/darkweird/Code/Terasology/gradle/gooeycli' is not part of the build defined by settings file '/home/darkweird/Code/Terasology/settings.gradle'. If this is an unrelated build, it must have its own settings file.
```

**Comment by DarkWeird** [comment]
image

Cannot initialize Omega

File: .idea/runConfigurations.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="RunConfigurationProducerService">
```
[the remark this answers is not in the document]
Comment (pollend) [reply .idea/runConfigurations.xml:3]: I need to delete that

**Comment by keturn** [comment]
If you've decided to stick with Groovy, you may be interested to know that gradle 7 ships with Groovy 3.0.

(Related: #4653)

**Comment by pollend** [comment]
> If you've decided to stick with Groovy, you may be interested to know that gradle 7 ships with Groovy 3.0.
> 
> (Related: #4653)

that sounds really nice

**Comment by DarkWeird** [comment]
We should restore `lib`/`meta`/`facade` commands
what else?

**Comment by pollend** [comment]
@DarkWeird I would like to umm, I don't think they were ever properly implement in the original tool.

**Comment by DarkWeird** [comment]
> @DarkWeird were those commands ever properly implemented?

Don't sure.
Groovyw just clone repos.
Haven't indexes.
And.. i didn't touch meta.

**Review by Cervator — CHANGES_REQUESTED** [review]
So this is definitely some solid work and I very much appreciate it, and am glad to have at least found enough time to go through the PR just about line by line. Nicely done 👍 

This feels like it does really well in areas where the work scratches an important itch, and we have several important itches. But it does also lose some things as a consequence and is somewhat incomplete. Additionally while my first stab was a definite draft effort that lacked stuff and included quirks, I think this one has some different quirks, particularly around some terminology. It also _really_ lacks better docs, be it in the javadoc, command descriptions, usage text, and PR description - even having made it through the whole thing I still don't really understand "snapshots", and that's _after_ having seen it get discussed on chat some and thusly having some more clues. It seems like an awesome feature we badly need, it just really needs some more polish and arranging things in a more user friendly fashion :-)

In that sense this reminds me of my own first effort, as both are incomplete. And I even have a _second_ incomplete effort that was a day job fork where I fixed some of the same issues this PR fixes, yet then I lack some things this covers and fixes other things this one doesn't. Ideally I'd find the time to try splicing my work improvements into this and wind up with a complete work we can then add a few more cool useful things to (like DW's tinkering).

One of the main things in my day job setup is that the workspace has an actual config file, which together with state helps figure out a bunch of things. It seems this mostly is a mix of hard coding and the constants class. Goal was to be able to ship something with workspace config ready out of the box, which in a work context might mean forks of a template that different teams customize. That might still fit in here with distros and temporary "snapshot" thing used to share workspace setup ("junctions" ?)

I don't know what sort of timeline I might be looking at, even reviewing this took all weekend and an innumerous amount of picking it up and putting it away again as I have to go pick up and put away tiny humans in RL. But I'm excited about it again at least so - maybe?

Thank you for the work @pollend (and @DarkWeird!) :-) To be continued, hopefully..

File: gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/workspace/snapshot/OpenCommand.groovy [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import java.awt.*
+
+@Command(name = "open", description = "open snapshot file")
+class OpenCommand implements Runnable {
+    @Parameters(paramLabel = "items", arity = "1", description = "tag snapshot with name")
+    String name = ""
+
+    @Override
+    void run() {
+        Desktop desktop = Desktop.getDesktop();
```
[the remark this answers is not in the document]
Comment (DarkWeird) [reply gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/workspace/snapshot/OpenCommand.groovy:19]: Yeah. 
Opens snapshot in your headed environment like double click at this file.

Don't works at headless :D

@pollend

File: gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/workspace/snapshot/LoadCommand.groovy [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Override
+    void run() {
+        File file = new File(target)
+        if (!file.exists()) {
+            return
+        }
+        Snapshot snapshot = new Snapshot(file)
+        snapshot.file = new File(Snapshot.SnapshotDirectory,"${snapshot.captured.epochSecond}.snapshot")
+        snapshot.save()
```
[the remark this answers is not in the document]
Comment (DarkWeird) [reply gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/workspace/snapshot/LoadCommand.groovy:25]: Load into local snapshot repo(just dir) from `external` locations

File: gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/module/GetCommand.groovy [hunk 7]
```diff
@@ -3,92 +3,77 @@
...
+import org.terasology.cli.module.ModuleIndex
+import org.terasology.cli.module.Modules
 import picocli.CommandLine
 import picocli.CommandLine.Command
-import picocli.CommandLine.Mixin
 import picocli.CommandLine.Option
 import picocli.CommandLine.Parameters
 
 @Command(name = "get", description = "get module ")
 class GetCommand implements Runnable {
```
Comment (Cervator) [inline gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/module/GetCommand.groovy:16]: So if I'm looking at things right there are two `GetCommand` classes, one common one module, and the one selected is simply based on the import statement? I guess that works but it looks scary to me :-)

If dependencies is the only unique way they differ, would it make sense to abstract out a `ProvidesDependencies` interface or something with a method, and if during the get command setup we see that only then look at the dependency stuff? Then you could in theory support dependencies per type arbitrarily

File: gradle/gooeycli/src/main/groovy/org/terasology/cli/options/GitOptions.groovy [hunk 8]
```diff
@@ -1,28 +1,32 @@
...
- * A mix-in meant to supply some general Git-related options
+ * A trait meant to supply some general Git-related options
  */
-class GitOptions {
-    @Option(names = ["-origin", "-o"], description = "Github remote Organization/User to target", defaultValue = Constants.DefaultModuleGithubOrg)
+trait GitOptions {
+
+    abstract ItemConfig getConfig()
+
+    @Option(names = ["-origin", "-o"], description = "Github remote Organization/User to target")
```
Comment (Cervator) [inline gradle/gooeycli/src/main/groovy/org/terasology/cli/options/GitOptions.groovy:16]: If we're doing Linux style should this become `--origin` ?

**Review by Cervator — COMMENTED** [review]
This looks very promising - I haven't tested so not approving just yet, not sure if I'll have enough time for that tonight. Thanks for the updates @DarkWeird 👍

**Comment by pollend** [comment]
I don't think we should be relying on "http://meta.terasology.org/modules/list/latest"

File: gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/module/GetCommand.groovy (hunk 7 above)
Comment (pollend) [reply gradle/gooeycli/src/main/groovy/org/terasology/cli/commands/module/GetCommand.groovy:17]: wonder what you gain by making this so abstract. trying to add additional layers to best fit the situation. wonder if it would just make more sense to implement this 3 or so times with some basic containers. feels a bit too clever for its own good.

**Comment by pollend** [comment]
I don't like how convoluted this is and trying to account for every possible edge case through some convoluted scheme of abstractions is kind of overkill. I would opt to shave a lot of this down to a single GitItem type and implement the details accordingly. with some helper utilities for some common things to share. this is a CLI utility, why does it need to be so complicated and a lot of this complicated logic feels unwarranted.

**Comment by pollend** [comment]
is there anything else?

**Comment by DarkWeird** [comment]
> is there anything else?

idk. :(
I haven't clean and full vision.

**Comment by skaldarnar** [comment]
I did have a brief look at this and played around with it for a bit. Here are my thoughts and notes:

- incomplete or buggy functionality
	- on running `gw module init iota` the following error occured:
	  ```
	  In copyInTemplateFiles for module ModuleTestingEnvironment - copying in a build.gradle then next checking for module.txt
	  Unable to clone ModuleTestingEnvironment, Skipping: Task java.util.concurrent.FutureTask@3a419fde[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@62ca1d89[Wrapped task = org.terasology.cli.commands.items.ModuleCommand$_recurseInner_closure3$_closure5@6626daa8]] rejected from java.util.concurrent.ThreadPoolExecutor@54906aa4[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 10] 
      ```
- document the new command line interface
  - update existing documentation mentioning `groovyw`
  - should we put explicit documentation about this tool somewhere?
  - explain how the whole groovyw project setup works?
  - not all commands and options have descriptions
- remove old `groovyw` code in a a separate PR
  - allows to work on docs in whatever time is needed
  - gives some time to keep both variants around to look out for issues
  - separate commit to remove the old code

Things I've tested:

- init a workspace with `gw module init iota`
- run a command in specific modules with `gw module cmd Health Drops Inventory -c='rm build.gradle'`
- refresh faulty modules with `gw module refresh`
- check the workspace status with `gw workspace status`
- update outdated modules with `gw module update`

A few other open questions:

- what can we do about this annoying "WARNING: An illegal reflective access operation has occurred"?
- running a simple `gw help` command takes 3 seconds
- output of `workspace status` is not completely clear to me, but usability can be improved in follow-ups

---

tl;dr: I'd like to split this PR up into smaller chunks in get it merged step by step.

**Comment by pollend** [comment]
> I did have a brief look at this and played around with it for a bit. Here are my thoughts and notes:
> 
>     * incomplete or buggy functionality
>       
>       * on running `gw module init iota` the following error occured:
>         ```
>         In copyInTemplateFiles for module ModuleTestingEnvironment - copying in a build.gradle then next checking for module.txt
>         Unable to clone ModuleTestingEnvironment, Skipping: Task java.util.concurrent.FutureTask@3a419fde[Not completed, task = java.util.concurrent.Executors$RunnableAdapter@62ca1d89[Wrapped task = org.terasology.cli.commands.items.ModuleCommand$_recurseInner_closure3$_closure5@6626daa8]] rejected from java.util.concurrent.ThreadPoolExecutor@54906aa4[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 10] 
>         ```
> 
>     * document the new command line interface
>       
>       * update existing documentation mentioning `groovyw`
>       * should we put explicit documentation about this tool somewhere?
>       * explain how the whole groovyw project setup works?
>       * not all commands and options have descriptions
> 
>     * remove old `groovyw` code in a a separate PR
>       
>       * allows to work on docs in whatever time is needed
>       * gives some time to keep both variants around to look out for issues
>       * separate commit to remove the old code
> 
> 
> Things I've tested:
> 
>     * init a workspace with `gw module init iota`
> 
>     * run a command in specific modules with `gw module cmd Health Drops Inventory -c='rm build.gradle'`
> 
>     * refresh faulty modules with `gw module refresh`
> 
>     * check the workspace status with `gw workspace status`
> 
>     * update outdated modules with `gw module update`
> 
> 
> A few other open questions:
> 
>     * what can we do about this annoying "WARNING: An illegal reflective access operation has occurred"?
> 
>     * running a simple `gw help` command takes 3 seconds
> 
>     * output of `workspace status` is not completely clear to me, but usability can be improved in follow-ups
> 
> 
> tl;dr: I'd like to split this PR up into smaller chunks in get it merged step by step.

there is no need to split the PR its self contained so it doesn't affect the existing cli tool.

**Comment by skaldarnar** [comment]
> [...] so it doesn't affect the existing cli tool.

Our [contributor guide](https://github.com/MovingBlocks/Terasology/wiki/Contributor-Quick-Start) asks the user to run 
```
groovyw module init iota
```
which does not work anymore with this PR. Thus, I don't follow the argument of this being "self-contained" and "not affecting the existing CLI tooling". 

In addition, as reported above, the replacement of exactly that comment threw an error when I tried it out.

**Comment by pollend** [comment]
> > [...] so it doesn't affect the existing cli tool.
> 
> Our [contributor guide](https://github.com/MovingBlocks/Terasology/wiki/Contributor-Quick-Start) asks the user to run
> 
> ```
> groovyw module init iota
> ```
> 
> which does not work anymore with this PR. Thus, I don't follow the argument of this being "self-contained" and "not affecting the existing CLI tooling".
> 
> In addition, as reported above, the replacement of exactly that comment threw an error when I tried it out.

roll back all the changes where I delete the old groovy code.
